stopper: wait on double Stop

knz opened the pull request:
Fixes #56728 

If there is another `Stop()` already in progress, make the second one
wait on the first one.

knz committed:
testcluster: properly wait for servers to terminate

Prior to this patch, a race condition existed in
`(*TestCluster).Stopper().Stop()` in case a test would also send a
`Drain()` RPC request to one of the individual servers with the
shutdown bit set: in that case, `Drain()` would initiate the stopper's
`Stop()` method, whereafter `(*TestCluster).Stopper()Stop()` would
take a shortcut and not actually wait on anything.

This patch fixes that by properly waiting.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/56729)
<!-- Reviewable:end -->

tbg commented on pkg/testutils/testcluster/testcluster.go:
```diff
@@ -108,6 +108,11 @@ func (tc *TestCluster) stopServers(ctx context.Context) {
+			// Stop() above to complete, because if there was an asynchronous
+			// Drain request ongoing, Stop() would take a shortcut and not
+			// actually wait for the async stop to complete.
+			<-tc.mu.serverStoppers[i].IsStopped()
```
`Stop()` should wait until `IsStopped()` is closed, though. I suggest adding a `<-s.IsStopped()` before the early return in `(*Stopper).Stop`.

knz replied:
Ok done. Let's see in CI whether there are tests that rely on this 2nd-stop-is-noop behavior.

tbg replied:
It is a noop :-) Just a bit slower now. I doubt anything will shake out but it's good to check of course.

knz committed:
stopper: wait on double Stop

If there is another `Stop()` already in progress, make the second one
wait on the first one.

Release note: None

knz reviewed (commented):
> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56729)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/testutils/testcluster/testcluster.go, line 115 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56729#-MMGKNu67wPyBawFpzlM-r1-115:-MMH-NJ80oWZFLx6YuoK:b-v96ohv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a1fe86030f90829c5be730be046371139513654/pkg/testutils/testcluster/testcluster.go#L115)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

It is a noop :-) Just a bit slower now. I doubt anything will shake out but it's good to check of course.
</blockquote></details>

i removed the code in the 2nd commit.


<!-- Sent from Reviewable.io -->

knz commented:
bors r=tbg
